docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,612 @@
+# Cross language object graph serialization
+
+Fury xlang serialization is an automatic object serialization framework that supports reference and polymorphism.
+Fury will convert an object from/to fury xlang serialization binary format.
+Fury has two core concepts for xlang serialization:
+
+- **Fury xlang binary format**
+- **Framework implemented in different languages to convert object to/from Fury xlang binary format**
+
+The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
+much more easy to use, but
+also introduce more complexities compared to static serialization frameworks. So the format will be more complex.
+
+## Type Systems
+
+### Data Types
+
+- bool: A boolean value (true or false).
+- byte: An 8-bit signed integer.
+- i16: A 16-bit signed integer.
+- i32: A 32-bit signed integer.
+- i64: A 64-bit signed integer.
+- half-float: A 16-bit floating point number.
+- float: A 32-bit floating point number.
+- double: A 64-bit floating point number including NaN and Infinity.
+- string: A text string encoded using Latin1/UTF16/UTF-8 encoding.
+- enum: a data type consisting of a set of named values. Rust enum with non-predefined field values are not supported as
+  an enum
+- list: A sequence of objects.
+- set: An unordered set of unique elements.
+- map: A map of key-value pairs.
```

Comment:
Is there any restriction ion the key type?

Reply:
Nope, it should be immutable, but this is nit related to the protocol

Reply:
I mean, what kind of types is allowed in key?

Reply:
list/map/set/array/tensor/arrow are not allowed, others are all allowed

Reply:
I think you can add it to the spec.

Reply:
updated